File: lintcode/python/0391_number_of_airplanes_in_the_sky.py
#!/usr/bin/python -t

# sweep line
# 使用九章算法强化班中讲过的扫描线算法

"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    """
    @param airplanes: An interval array
    @return: Count of airplanes are in the sky.
    """
    def countOfAirplanes(self, airplanes):
        # write your code here
        cnts = []
        
        for airplane in airplanes:
            cnts.append([airplane.start, 1])
            cnts.append([airplane.end, 0])
            
        ret = cur = 0
        
        for _, start in sorted(cnts):
            if start == 1:
                cur += 1
            else:
                cur -= 1
                
            ret = max(ret, cur)
            
        return ret

File: lintcode/golang/0391_number_of_airplanes_in_the_sky.go
package main

//sweep line

// time O(nlogn) due to the sort
// space O(n)

import (
	"fmt"
	"sort"
)

type Interval struct {
	Start, End int
}

type point struct {
	time int
	cnt  int
}

/**
 * @param airplanes: An interval array
 * @return: Count of airplanes are in the sky.
 */
func countOfAirplanes(airplanes []*Interval) int {
	// write your code here
	n := len(airplanes)
	if n == 0 {
		return 0
	}

	points := make([]*point, 0)

	for _, a := range airplanes {
		points = append(points, &point{time: a.Start, cnt: 1})
		points = append(points, &point{time: a.End, cnt: -1})
	}

	sort.Slice(points, func(i, j int) bool {
		return points[i].time < points[j].time ||
			(points[i].time == points[j].time && points[i].cnt < points[j].cnt)
	})

	ret := 0
	cur := 0

	for _, p := range points {
		cur += p.cnt
		ret = max(ret, cur)
	}

	return ret
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	a := []*Interval{
		&Interval{1, 10},
		&Interval{2, 3},
		&Interval{5, 8},
		&Interval{4, 7},
	}
	fmt.Println(countOfAirplanes(a))
}
